Release test for AWS IAM Redis passwordless

raviharshicorp opened the pull request:
These changes enable TFE to use AWS IAM authentication for Redis connections instead of password-based authentication.

## Background
Added a spec for testing AWS redis passwordless


Part of AWS redis passwordless release test
Related PRs
terraform-enterprise # https://github.com/hashicorp/terraform-enterprise/pull/3149
terraform-terraform-enterprise-fdo: https://github.com/hashicorp/terraform-terraform-enterprise-fdo/pull/67
ptfedev-infra: https://github.com/hashicorp/ptfedev-infra/pull/885


## How has this been tested?
CI/CD: https://github.com/hashicorp/terraform-enterprise/actions/runs/19707496447/job/56458820019


[JIRA](https://hashicorp.atlassian.net/browse/IND-5407)

## Screenshots
<img width="1512" height="780" alt="Screenshot 2025-11-27 at 11 01 59" src="https://github.com/user-attachments/assets/afef43ac-7e22-4d4c-8257-177a9cffac9e" />

<img width="898" height="462" alt="Screenshot 2025-11-20 at 14 04 31" src="https://github.com/user-attachments/assets/6552146c-9518-440b-ab73-1c41f3dd2146" />

<img width="1512" height="309" alt="Screenshot 2025-11-20 at 14 09 42" src="https://github.com/user-attachments/assets/a76c007e-0ab2-4be1-b6a3-32f8b53f4e45" />

<img width="1512" height="319" alt="Screenshot 2025-11-20 at 14 10 06" src="https://github.com/user-attachments/assets/2e9538dd-3e64-4b8f-adc5-096107d4bbec" />

raviharshicorp committed:
Add AWS IAM Redis passwordless authentication variables

- Add redis_passwordless_aws_use_iam variable to enable AWS IAM Redis auth
- Add redis_passwordless_aws_region variable for region specification
- Configure TFE_REDIS_PASSWORDLESS_AWS_USE_IAM environment variable
- Configure TFE_REDIS_SIDEKIQ_PASSWORDLESS_AWS_USE_IAM for Sidekiq
- Configure TFE_REDIS_PASSWORDLESS_AWS_REGION environment variable

These changes enable TFE to use AWS IAM authentication for Redis connections
instead of password-based authentication, following the same pattern as
Azure MSI Redis authentication.

raviharshicorp committed:
feat: add Sidekiq Redis connection support for passwordless auth

- Add TFE_REDIS_SIDEKIQ_* environment variables for separate Sidekiq Redis
- Add TFE_REDIS_SIDEKIQ_PASSWORDLESS_AWS_* variables for IAM authentication
- Add TFE_REDIS_PASSWORDLESS_AWS_HOST_NAME for cluster name extraction
- Support separate Redis instances for main and Sidekiq with fallback to main Redis
- Add variables: redis_sidekiq_host, redis_sidekiq_user, redis_sidekiq_password
- Enable dual Redis passwordless authentication (main + Sidekiq)

raviharshicorp committed:
debug: Add Redis environment variables debugging outputs

- Add debug output for all Redis environment variables
- Add debug output for input variables received
- Track Redis username propagation to TFE container

raviharshicorp committed:
fix: Add missing AWS IAM Redis passwordless authentication variables

- Add redis_passwordless_aws_use_iam variable to variables.tf
- Add TFE_REDIS_PASSWORDLESS_AWS_USE_IAM and TFE_REDIS_SIDEKIQ_PASSWORDLESS_AWS_USE_IAM environment variables to redis_config.tf
- Add debug outputs for Redis environment variables and input variables
- Ensures Redis IAM authentication is properly configured for AWS ElastiCache

raviharshicorp committed:
Merge branch 'main' into pravi/IND-5861

raviharshicorp committed:
fix: Apply terraform fmt to debug outputs in runtime_container_engine_config

- Fix alignment and indentation in debug_redis_env_vars and debug_redis_input_vars outputs
- Standardize spacing to align equals signs consistently with terraform formatting standards
- Resolves terraform fmt check failure in CI pipeline

raviharshicorp committed:
Merge branch 'pravi/IND-5861' of github.com:hashicorp/terraform-random-tfe-utility into pravi/IND-5861

ajmera-naman commented on modules/runtime_container_engine_config/outputs.tf:
```diff
@@ -10,3 +10,26 @@ output "podman_kube_yaml" {
 }
+
+output "debug_redis_env_vars" {
+  value = {
```
we can remove this debug outpus

raviharshicorp replied:
@ajmera-naman removed all debug outputs

ajmera-naman commented on modules/runtime_container_engine_config/outputs.tf:
```diff
@@ -10,3 +10,26 @@ output "podman_kube_yaml" {
+  description = "Debug output for Redis environment variables"
+}
+
+output "debug_redis_input_vars" {
```
same as above

raviharshicorp replied:
@ajmera-naman removed all debug outputs

ajmera-naman commented on modules/runtime_container_engine_config/redis_config.tf:
```diff
@@ -20,6 +20,8 @@ locals {
     TFE_REDIS_PASSWORDLESS_AZURE_USE_MSI         = var.redis_passwordless_azure_use_msi
     TFE_REDIS_SIDEKIQ_PASSWORDLESS_AZURE_USE_MSI = var.redis_passwordless_azure_use_msi
     TFE_REDIS_PASSWORDLESS_AZURE_CLIENT_ID       = var.redis_passwordless_azure_client_id
+    TFE_REDIS_PASSWORDLESS_AWS_USE_IAM           = var.redis_passwordless_aws_use_iam
```
This is incorrect
config values - https://developer.hashicorp.com/terraform/enterprise/deploy/reference/configuration#tfe_redis_passwordless_aws_use_instance_profile

ajmera-naman replied:
@raviharshicorp the configuration is still incorrect, pls look at the attached link

raviharshicorp replied:
@ajmera-naman Corrected

ajmera-naman reviewed (changes requested):
Configuration seems incorrect

kkavish commented on modules/runtime_container_engine_config/redis_config.tf:
```diff
@@ -20,6 +20,8 @@ locals {
     TFE_REDIS_SIDEKIQ_PASSWORDLESS_AZURE_USE_MSI = var.redis_passwordless_azure_use_msi
     TFE_REDIS_PASSWORDLESS_AZURE_CLIENT_ID       = var.redis_passwordless_azure_client_id
+    TFE_REDIS_PASSWORDLESS_AWS_USE_IAM           = var.redis_passwordless_aws_use_iam
+    TFE_REDIS_SIDEKIQ_PASSWORDLESS_AWS_USE_IAM   = var.redis_passwordless_aws_use_iam
```
same as @ajmera-naman mentioned above, the problem I feel now is deeper. If the configs are incorrect, the Passwordless auth shouldn't work at all. These are either working with standard auth or we are looking at the incorrect test case.

raviharshicorp replied:
@kkavish Updated configuration with correct variables.

raviharshicorp committed:
fix: correct Redis AWS IAM variable names and add database AWS IAM support

- Change redis_passwordless_aws_use_iam to redis_passwordless_aws_use_instance_profile
- Add database_passwordless_aws_use_instance_profile and database_passwordless_aws_region variables
- Remove debug outputs as requested in PR review
- Variables now match TFE documentation exactly

Addresses PR review comments:
- Configuration now uses correct TFE environment variable names
- Database AWS IAM authentication support added
- Debug outputs removed for cleaner module interface

raviharshicorp committed:
cleanup: remove unnecessary PostgreSQL AWS IAM variables

- Remove database_passwordless_aws_use_instance_profile variable
- Remove database_passwordless_aws_region variable
- Remove TFE_DATABASE_PASSWORDLESS_AWS_* environment variables

PostgreSQL AWS IAM authentication is out of scope for this Redis
passwordless authentication feature. These variables were incorrectly
added during the variable name fix and should not be included.

Focus remains purely on Redis passwordless authentication with correct
variable names matching TFE documentation.

raviharshicorp committed:
Add missing Redis passwordless AWS variables and TFE environment variables

- Add redis_passwordless_aws_region and redis_passwordless_aws_host_name variables
- Add corresponding TFE environment variables for Redis AWS region and hostname
- Add Sidekiq variants for both region and hostname
- Apply terraform formatting alignment for consistency
- Complete Redis passwordless authentication implementation

raviharshicorp committed:
Add missing TFE Redis Sidekiq environment variables per documentation

- Added TFE_REDIS_SIDEKIQ_USER (set to redis_user for IAM authentication)
- Added TFE_REDIS_SIDEKIQ_USE_TLS (set to redis_use_tls value)

These variables are required by the official TFE documentation for Redis IAM authentication.

raviharshicorp committed:
Fix Redis username and password for IAM authentication

- Use 'default' username when redis_passwordless_aws_use_instance_profile=true
- Set password to null for IAM authentication (TFE generates tokens dynamically)
- Apply same fix for both TFE_REDIS_USER and TFE_REDIS_SIDEKIQ_USER

raviharshicorp committed:
Update Redis config to use custom IAM user for testing

- Use redis_passwordless_aws_iam_user instead of hardcoded 'default'
- Add redis_passwordless_aws_iam_user variable
- Apply to both TFE_REDIS_USER and TFE_REDIS_SIDEKIQ_USER

This enables proper testing of custom ElastiCache IAM users per AWS documentation
instead of relying on the 'default' user which bypasses IAM validation.

raviharshicorp committed:
Temporarily revert to 'default' user to test Redis IAM auth works

This will confirm that:
1. Redis IAM authentication mechanism is working
2. The issue is specifically with custom IAM user configuration

Once confirmed working, we'll know the problem is the ElastiCache
user group doesn't properly include the custom IAM user 'fitg-iam-user'.

raviharshicorp committed:
Revert Redis config to use custom IAM user

Now that IAM policy is fixed with explicit resource ARNs,
we can test the custom IAM user authentication properly.

ajmera-naman reviewed (approved):
LGTM

nikolasrieble commented on modules/runtime_container_engine_config/redis_config.tf:
```diff
@@ -3,23 +3,31 @@
-    TFE_REDIS_PASSWORDLESS_AZURE_CLIENT_ID       = var.redis_passwordless_azure_client_id
+    TFE_REDIS_HOST                                          = var.redis_use_tls != null ? var.redis_use_tls ? "${var.redis_host}:6380" : var.redis_host : null
+    TFE_REDIS_USER                                          = var.redis_passwordless_aws_use_instance_profile ? var.redis_passwordless_aws_iam_user : var.redis_user
+    TFE_REDIS_PASSWORD                                      = var.redis_passwordless_aws_use_instance_profile ? null : var.redis_password
```
It might be nice to group all items affected by the var.redis_passwordless_aws_use_instance_profile by moving the lines closer to each other.

raviharshicorp replied:
@nikolasrieble Updated.

nikolasrieble commented on modules/runtime_container_engine_config/redis_config.tf:
```diff
@@ -3,23 +3,31 @@
+    TFE_REDIS_SIDEKIQ_PASSWORDLESS_AWS_REGION               = var.redis_passwordless_aws_region
+    TFE_REDIS_PASSWORDLESS_AWS_HOST_NAME                    = var.redis_passwordless_aws_host_name
+    TFE_REDIS_SIDEKIQ_PASSWORDLESS_AWS_HOST_NAME            = var.redis_passwordless_aws_host_name
+    TFE_REDIS_SIDEKIQ_USER                                  = var.redis_passwordless_aws_use_instance_profile ? var.redis_passwordless_aws_iam_user : var.redis_user
```
Great catch! Why is this change included in the PR? 

❓ How is it possible that our release tests for `redis_use_tls` ever passed ✅, if we never passed this configuration here?

raviharshicorp replied:
The TFE_REDIS_SIDEKIQ_USE_TLS variable was indeed missing from the configuration before.
While implementing Redis passwordless authentication, I noticed that Sidekiq-specific Redis configuration was incomplete. TFE has separate Redis connections for main application and Sidekiq background jobs - Sidekiq wasn't getting explicit TLS configuration. Sidekiq likely has application-level fallback logic that inherits TLS settings when not explicitly configured.

 it was discovered during Redis IAM work and both changes touch Redis config, I included it here with proper documentation explaining it's unrelated to IAM auth.

nikolasrieble commented on modules/runtime_container_engine_config/variables.tf:
```diff
@@ -369,6 +369,30 @@ variable "redis_passwordless_azure_client_id" {
+  default     = false
+}
+
+variable "redis_passwordless_aws_region" {
+  type        = string
+  description = "AWS Region of the AWS ElastiCache resource for Redis passwordless authentication."
+  default     = null
+}
+
+variable "redis_passwordless_aws_host_name" {
+  type        = string
+  description = "The name of the Redis instance on AWS for passwordless authentication."
+  default     = null
+}
+
+variable "redis_passwordless_aws_iam_user" {
+  type        = string
+  description = "The IAM username for Redis IAM authentication."
+  default     = null
+}
```
❓ Would it make sense to group these instead? 

```
variable "redis_passwordless_config" {
  type = object({
    aws_region  = string
    host_name   = string
    iam_user    = string
  })
  description = "Configuration for Redis passwordless authentication."
}
```

That might be a better user experience, and communicate that they should be set together.

raviharshicorp replied:
This would break existing module usage where these variables are already consumed.

nikolasrieble reviewed (commented):
The changes are fine, I have only minor comments to consider. 

One thing thought deserve an independent follow up: This PR sets the unrelated var `TFE_REDIS_SIDEKIQ_USE_TLS`.

How come we missed this in the past? Do we have release tests for the USE_TLS feature? If yes, why would they ever pass, if we did not pass on this flag to SideKiq?

raviharshicorp committed:
Extract repeated redis_passwordless_aws_use_instance_profile condition into local variable

Address feedback from nikolasrieble and GitHub Copilot:
- Extract var.redis_passwordless_aws_use_instance_profile into local.use_redis_passwordless_aws
- Replace 5 occurrences of the repeated condition with the local variable
- Improves maintainability and reduces duplication as suggested

This makes the code easier to maintain if the condition logic ever needs to change.

raviharshicorp committed:
Add documentation for TFE_REDIS_SIDEKIQ_USE_TLS variable

Address Copilot feedback by adding comments explaining:
- TFE_REDIS_SIDEKIQ_USE_TLS ensures Sidekiq uses TLS when required
- This variable fixes missing Sidekiq TLS configuration discovered during IAM auth work
- Addition is unrelated to AWS IAM authentication functionality

This provides context for why Sidekiq-specific TLS configuration was added.

ajmera-naman commented on modules/runtime_container_engine_config/redis_config.tf:
```diff
@@ -2,24 +2,37 @@
-    TFE_REDIS_SIDEKIQ_PASSWORDLESS_AZURE_USE_MSI = var.redis_passwordless_azure_use_msi
-    TFE_REDIS_PASSWORDLESS_AZURE_CLIENT_ID       = var.redis_passwordless_azure_client_id
+    TFE_REDIS_HOST                                          = var.redis_use_tls != null ? var.redis_use_tls ? "${var.redis_host}:6380" : var.redis_host : null
+    TFE_REDIS_USER                                          = local.use_redis_passwordless_aws ? var.redis_passwordless_aws_iam_user : var.redis_user
```
I don't see the variable `redis_passwordless_aws_iam_user` coming from the terraform-fdo repo

raviharshicorp commented:
Closing this as this changes in this repo is no longer relevant.
changes are ported here: https://github.com/hashicorp/terraform-terraform-enterprise-fdo/pull/67/files
